packages/lodestar-beacon-state-transition/src/fast/block/initiateValidatorExit.ts

```diff
@@ -29,3 +32,40 @@ export function initiateValidatorExit(epochCtx: EpochContext, state: BeaconState
   validator.exitEpoch = exitQueueEpoch;
   validator.withdrawableEpoch = validator.exitEpoch + config.params.MIN_VALIDATOR_WITHDRAWABILITY_DELAY;
 }
+
+/**
+ * Optimized version of initiateValidatorExit where we process validators in batch.
+ * The main thing is to loop through state.validators exactly 1 time.
+ */
+export function initiateMultipleValidatorExits(
+  epochCtx: EpochContext,
+  state: BeaconState,
+  indexes: ValidatorIndex[] = []
+): void {
+  if (!indexes || indexes.length === 0) return;
+  const config = epochCtx.config;
+  // compute exit queue epoch
+  const validatorExitEpochs = readOnlyMap(state.validators, (v) => v.exitEpoch);
+  const exitEpochs = validatorExitEpochs.filter((exitEpoch) => exitEpoch !== FAR_FUTURE_EPOCH);
```

Comment:
would it be better to filter first then map?

Reply:
Definitely! Also, why not do both in one loop?

Reply:
Can't in a single loop unless we add filerAndmapReadOnly iterator to ssz. I hope we have readOnlyFilter if not, this mihgt be useless comment

Reply:
Can you do this?
```ts
const exitEpochs = []
for (const v of state.validators) {
  if (v.exitEpoch !== FAR_FUTURE_EPOCH) {
    exitEpochs.push(v.exitEpoch)
  }
}
```

Reply:
Keep in mind that state is actually TreeBacked so if I'm not mistaken, your code would actually deserialize and allocate new validator objects. But your idea could be done with readOnlyForEach. Ref: https://github.com/ChainSafe/ssz/blob/master/src/backings/tree/array.ts#L141

Reply:
`readOnlyForEach` would be good, will do it